### PR TITLE
feat:implement sql style 'substr_index' string function

### PR DESCRIPTION
**Pull request by Syleechan**
## Which issue does this PR close?

Closes #.

## Rationale for this change
https://dev.mysql.com/doc/refman/8.0/en/string-functions.html#function_substring-index:~:text=SUBSTRING_INDEX(str%2Cdelim%2Ccount)
In calcite, the function expression name is substr_index

### CODE REVIEW

File: datafusion/sqllogictest/test_files/functions.slt [hunk 1]
```diff
@@ -877,3 +877,63 @@ query ?
...
+SELECT substr_index('www.apache.org', NULL, 1)
+----
+NULL
+
+query T
+SELECT substr_index('www.apache.org', 'ac', NULL)
+----
+NULL
+
+query ?
```
Comment (comphead) [inline datafusion/sqllogictest/test_files/functions.slt:936]: awesome, can we also have the same tests with empty strings as input and search token?

File: datafusion/physical-expr/src/unicode_expressions.rs [hunk 2]
```diff
@@ -455,3 +455,61 @@ pub fn translate<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
...
     Ok(Arc::new(result) as ArrayRef)
 }
+
+/// Returns the substring from str before count occurrences of the delimiter delim. If count is positive, everything to the left of the final delimiter (counting from the left) is returned. If count is negative, everything to the right of the final delimiter (counting from the right) is returned.
+/// SUBSTRING_INDEX('www.apache.org', '.', 1) = www
+/// SUBSTRING_INDEX('www.apache.org', '.', 2) = www.apache
+/// SUBSTRING_INDEX('www.apache.org', '.', -2) = apache.org
+/// SUBSTRING_INDEX('www.apache.org', '.', -1) = org
+pub fn substr_index<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
+    let string_array = as_generic_string_array::<T>(&args[0])?;
```
Comment (comphead) [inline datafusion/physical-expr/src/unicode_expressions.rs:465]: we need to add a defense check args is exactly 3 elements

**Review by comphead — COMMENTED** [review]
Thanks @Syleechan please take a look into minor comments

File: datafusion/sqllogictest/test_files/functions.slt (hunk 1 above)
Comment (Syleechan) [reply datafusion/sqllogictest/test_files/functions.slt:936]: > awesome, can we also have the same tests with empty strings as input and search token?

maybe the line 921 ~ 924 meet your expect test?

File: datafusion/physical-expr/src/unicode_expressions.rs (hunk 2 above)
Comment (Syleechan) [reply datafusion/physical-expr/src/unicode_expressions.rs:472]: thanks, I add the args len check

File: datafusion/sqllogictest/test_files/functions.slt (hunk 1 above)
Comment (Syleechan) [reply datafusion/sqllogictest/test_files/functions.slt:951]: add empty string tests and 0 count tests

**Review by Ted-Jiang — APPROVED** [review]
LGTM 👍 , and  tested below  in mysql, maybe add them is better
`
SELECT SUBSTRING_INDEX('www.mysql.com', '.', 4);  -> www.mysql.com
SELECT SUBSTRING_INDEX('www.mysql.com', '.', 0);  ->  
`

**Comment by Syleechan** [comment]
> LGTM 👍 , and tested below in mysql, maybe add them is better `SELECT SUBSTRING_INDEX('www.mysql.com', '.', 4); -> www.mysql.com SELECT SUBSTRING_INDEX('www.mysql.com', '.', 0); -> `

**Comment by Syleechan** [comment]
> LGTM 👍 , and tested below in mysql, maybe add them is better `SELECT SUBSTRING_INDEX('www.mysql.com', '.', 4); -> www.mysql.com SELECT SUBSTRING_INDEX('www.mysql.com', '.', 0); -> `

just the same as SELECT substr_index('www.apache.org', 'ac', 2)

File: datafusion/physical-expr/src/unicode_expressions.rs [hunk 3]
```diff
@@ -455,3 +455,68 @@ pub fn translate<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
...
 }
+
+/// Returns the substring from str before count occurrences of the delimiter delim. If count is positive, everything to the left of the final delimiter (counting from the left) is returned. If count is negative, everything to the right of the final delimiter (counting from the right) is returned.
+/// SUBSTRING_INDEX('www.apache.org', '.', 1) = www
+/// SUBSTRING_INDEX('www.apache.org', '.', 2) = www.apache
+/// SUBSTRING_INDEX('www.apache.org', '.', -2) = apache.org
+/// SUBSTRING_INDEX('www.apache.org', '.', -1) = org
+pub fn substr_index<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
+    if args.len() != 3 {
+        return Err(DataFusionError::Internal(format!(
```
Comment (comphead) [inline datafusion/physical-expr/src/unicode_expressions.rs:466]: please use `internal_err!` macros

Comment (Syleechan) [reply datafusion/physical-expr/src/unicode_expressions.rs]: thanks, change to internal_err.

**Comment by Syleechan** [comment]
@alamb please help to merge if there are no other concerns, thanks.

**Review by comphead — APPROVED** [review]
lgtm thanks @Syleechan

**Comment by alamb** [comment]
Thanks @Syleechan as well as @comphead  and @Ted-Jiang  for the review

Since @Ted-Jiang  is a committer as well, in the future please feel free to merge PRs once they have been reviewed and follow https://arrow.apache.org/datafusion/contributor-guide/index.html#pull-request-overview (as this one has)

**Pull request merged**